toolchain/lex/string_literal.cpp

```diff
@@ -341,8 +353,8 @@ static auto ExpandAndConsumeEscapeSequence(LexerDiagnosticEmitter& emitter,
 // Expand any escape sequences in the given string literal.
 static auto ExpandEscapeSequencesAndRemoveIndent(
     LexerDiagnosticEmitter& emitter, llvm::StringRef contents, int hash_level,
-    llvm::StringRef indent) -> std::string {
-  std::string result;
+    llvm::StringRef indent) -> GeneratedString {
+  GeneratedString result;
   result.reserve(contents.size());
```

Comment:
Hm! It strikes me that the result will *always* fit into this reserved storage, because escape sequences always expand to something no longer than themselves (in fact, always at least one character shorter currently). So instead of using a `SmallString`, we could allocate `contents.size()` from the `BumpPtrAllocator`, and fill it directly. That should get us to a single allocation and no copies.

(We may have a little unused space at the end of the allocation, which we could return to the `BumpPtrAllocator` if it had an API to do that, but currently no such API exists.)

Reply:
Done. Let me know if you see an issue, but I've done my best to translate the code and the performance still isn't good. Still, maybe it's easier to improve like this.

```
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
BM_ValidString_Simple                                  31166 ns        31166 ns        22384
BM_ValidString_Multiline                               32121 ns        32119 ns        21954
BM_ValidString_MultilineDoubleQuote                    31869 ns        31863 ns        21758
BM_ValidString_Raw                                     31928 ns        31928 ns        22025
BM_IncompleteWithEscapes_Simple                        47904 ns        47893 ns        13754
BM_IncompleteWithEscapes_Multiline                     41636 ns        41636 ns        16672
BM_IncompleteWithEscapes_MultilineDoubleQuote          42071 ns        42058 ns        16703
BM_IncompleteWithEscapes_Raw                          104998 ns       104988 ns         6697
BM_ComputeValue_NoGenerate_Short                        20.7 ns         20.7 ns     33566715
BM_ComputeValue_NoGenerate_Long                         3164 ns         3163 ns       220592
BM_ComputeValue_WillGenerate_Short                      52.0 ns         52.0 ns     13129991
BM_ComputeValue_WillGenerate_Long                      16368 ns        16366 ns        44319
BM_ComputeValue_WillGenerate_Multiline                 16194 ns        16191 ns        42727
BM_ComputeValue_WillGenerate_MultilineDoubleQuote      16438 ns        16435 ns        43282
BM_ComputeValue_WillGenerate_Raw                       16286 ns        16279 ns        43080
```

Reply:
I can reproduce your results. Here's what I think:

- The slowdown in `IncompleteWithEscapes` is due to computing the `content_needs_validation` flag. I think something weird is happening when optimizing `StringLiteral::Lex`. Making the flag `volatile` helps substantially.
- The slowdown in `ComputeValue` is a measuring artifact. You're hitting the best case for the old code (repeatedly deallocating and reallocating a heap region of the exact same size over and over again) and hitting a bad / typical case for the new code (performing a long series of allocations into the same bump pointer allocator). Most of the difference goes away if you call `Reset` on the `BumpPtrAllocator` each loop iteration, but some difference remains.

Reply:
Having spent a while looking at this with @chandlerc, the differences in at least `ValidString` and `IncompleteWithEscapes` seem to be performance ghosts caused by CPU weirdnesses like function alignment rather than an actual performance difference in the code. And the slowdown on the `ComputeValue` benchmark seems to be a measuring artifact. So I think the conclusion is that we're fine from a performance perspective. :shrug:

Reply:
For posterity, I think a fairer way to compare the results here is to add to `BM_ComputeValue` a `std::vector<std::string>` to hold the results in the cases without a `BumpPtrAllocator`, so that in all cases we store all allocations until the end of the benchmark. With that, I get:

Baseline (+ `std::vector<std::string>` non-stable storage):
```
BM_ComputeValue_NoGenerate_Short                        67.7 ns         67.7 ns     11734809
BM_ComputeValue_NoGenerate_Long                        15299 ns        15298 ns        46581
BM_ComputeValue_WillGenerate_Short                      89.5 ns         89.5 ns      9173553
BM_ComputeValue_WillGenerate_Long                      15293 ns        15292 ns        46916
BM_ComputeValue_WillGenerate_Multiline                 16082 ns        16082 ns        43286
BM_ComputeValue_WillGenerate_MultilineDoubleQuote      15793 ns        15790 ns        45479
BM_ComputeValue_WillGenerate_Raw                       15544 ns        15541 ns        45996
```

With the `std::unique_ptr<std::string>` version (+ `std::vector<StringLiteral::ComputedValue>` stable storage):
```
BM_ComputeValue_NoGenerate_Short                        56.0 ns         56.0 ns     10000000
BM_ComputeValue_NoGenerate_Long                         4770 ns         4770 ns       145437
BM_ComputeValue_WillGenerate_Short                       116 ns          116 ns      6843959
BM_ComputeValue_WillGenerate_Long                      15492 ns        15492 ns        44583
BM_ComputeValue_WillGenerate_Multiline                 15527 ns        15526 ns        43523
BM_ComputeValue_WillGenerate_MultilineDoubleQuote      15569 ns        15568 ns        43896
BM_ComputeValue_WillGenerate_Raw                       16037 ns        16036 ns        42307
```

This PR (with `llvm::BumpPtrAllocator` for stable storage):
```
BM_ComputeValue_NoGenerate_Short                        24.2 ns         24.1 ns     29077659
BM_ComputeValue_NoGenerate_Long                         4725 ns         4724 ns       147751
BM_ComputeValue_WillGenerate_Short                      62.9 ns         62.9 ns     11202462
BM_ComputeValue_WillGenerate_Long                      17900 ns        17898 ns        39221
BM_ComputeValue_WillGenerate_Multiline                 17984 ns        17981 ns        39195
BM_ComputeValue_WillGenerate_MultilineDoubleQuote      17886 ns        17885 ns        39069
BM_ComputeValue_WillGenerate_Raw                       17952 ns        17949 ns        38389
```

I think the first three of these are probably the most common and most important, and this approach seems to win out there. The bottom four cases exceed the `BumpPtrAllocator`'s default slab size, so they're allocating custom slabs, which is a little less efficient than doing a direct heap allocation here because the `BumpPtrAllocator` adds tracking overhead. So I think these numbers make sense.